source/features/show-whitespaces.tsx

```diff
@@ -0,0 +1,99 @@
+import './show-whitespaces.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import onPrFileLoad from '../libs/on-pr-file-load';
+import onNewComments from '../libs/on-new-comments';
+
+const queue: Node[] = [];
+
+// Process a single line for each frame loop
+function loop(): void {
+	const line = queue.shift();
+
+	if (line === undefined) {
+		return;
+	}
+
+	const iterator = document.createNodeIterator(line, NodeFilter.SHOW_TEXT, {
+		acceptNode: node => {
+			if (node.childNodes.length === 0) {
+				return NodeFilter.FILTER_ACCEPT;
+			}
+
+			return NodeFilter.FILTER_REJECT;
+		}
+	});
+
+	const textNodes = [];
+	let node;
+
+	// Collect all nodes before modifying the root node anymore
+	while ((node = iterator.nextNode())) {
+		textNodes.push(node);
+	}
+
+	for (const textNode of textNodes) {
+		const nodeValue = textNode.textContent!;
+		if (nodeValue.length !== 0) {
+			const fragment = document.createDocumentFragment();
+
+			for (const char of nodeValue) {
+				if (char === '\t') {
+					fragment.append(<span className="pl-ws pl-tab">{char}</span>);
+				} else if (char === ' ') {
+					fragment.append(<span className="pl-ws pl-space">{char}</span>);
+				} else {
+					fragment.append(char);
+				}
+			}
+
+			(textNode as Element).replaceWith(fragment);
+		}
+	}
+
+	// Add a new-line character at the end (optional)
+	// const br = line.querySelector('br');
+	// line.insertBefore(<span className="pl-ws pl-nl">&nbsp;</span>, br);
+
+	requestAnimationFrame(loop);
```

Comment:
This is processing one line per frame, not very efficient. If you want to split up the job without blocking the rendering, process X lines at a time. I think that if you use many `setTimeout`s the rendered will do as much as possible in the current frame and then continue in the next one.

But for now this sounds like premature optimization, let's see how bad it performs first.

Reply:
How about processing one table for each frame, that should be a reasonable compromise.

Reply:
This is not heavy processing, plus if we wait for frames the code won’t run until the page is visible again

Reply:
It did become a heavy task causing the page to jank on load on very large PR like the Parcel example above. After which I moved to `rAF`.

If only we had a way to execute a function repeatedly for specific duration and schedule the next call when the main thread is free (chunking of work).

I know that `requestIdleCallback` exists but it does work a lot similar to `rAF` in this case, and also has no chunking of work.

Reply:
If the attribute-based suggestion works it will become much lighter though